docs/Customization/Custom-Core.rst

```diff
@@ -0,0 +1,404 @@
+support AHB, APB and AXIS, and most of the AXI4 widgets has equivalent widget for these bus protocol. See the reference page for
+more info. 
+
+``memAXI4Node`` is an AXI4 master node and is defined as following in our example:
+
+.. code-block:: scala
+
+    val memAXI4Node = AXI4MasterNode(
+    Seq(AXI4MasterPortParameters(
+      masters = Seq(AXI4MasterParameters(
+        name = portName,
+        id = IdRange(0, 1 << idBits))))))
+
+where ``portName`` and ``idBits`` (number of bits to represent a port ID) are the parameter provides by the tile.
+Make sure to read :ref:`node_types` to check out what type of nodes Chipyard supports and their parameters!
+
+Also, by default, there are boundary buffers for both master and slave connections to the bus when they are leaving the tile, and you
+can override the following two functions to control how to buffer the bus requests/responses:
+
+.. code-block:: scala
+
+    protected def makeMasterBoundaryBuffers(implicit p: Parameters): TLBuffer
+    protected def makeSlaveBoundaryBuffers(implicit p: Parameters): TLBuffer
+
+You can find more information on ``TLBuffer`` in :ref:`diplomatic_widgets`.
+
+Interrupt
+---------
+
+Chipyard allows a tile to either receive interrupts from other devices or initiate interrupts to notify other cores/devices. 
+In the tile that inherited ``SinksExternalInterrupts``, one can create a ``TileInterrupts`` object (a Chisel bundle) and 
+call ``decodeCoreInterrupts`` with the object as the argument. You can then read the interrupt bits from the object.
+The definition of ``TileInterrupts`` is 
+
+.. code-block:: scala
+
+    class TileInterrupts(implicit p: Parameters) extends CoreBundle()(p) {
+      val debug = Bool() // debug interrupt
+      val mtip = Bool() // Machine level timer interrupt
+      val msip = Bool() // Machine level software interrupt
+      val meip = Bool() // Machine level external interrupt 
+      val seip = usingSupervisor.option(Bool()) // Valid only if supervisor mode is supported
+      val lip = Vec(coreParams.nLocalInterrupts, Bool())  // Local interrupts
+    }
+
+This function should be in the implementation class since it involves hardware generation. 
+Also, the tile can also notify other cores or devices for some events by calling following functions (in implementation class):
+
+.. code-block:: scala
+
+    def reportHalt(could_halt: Option[Bool]) // Triggered when there is an unrecoverable hardware error (halt the machine)
+    def reportHalt(errors: Seq[CanHaveErrors]) // Varient for standard error bundle (used only by cache when there's an ECC error)
+    reportCease(could_cease: Option[Bool], quiescenceCycles: Int = 8) // Triggered when the core stop retiring instructions (like clock gating)
+    reportWFI(could_wfi: Option[Bool]) // Triggered when a WFI instruction is executed
+
+Trace (Optional)
+----------------
+
```

Comment:
I think we should leave this section out for now. We actually use a nonstandard trace format for debugging, since its easier to use than the official Trace spec.

Reply:
Done. I just removed this part.